foomatic-db ppd files

Yarny0 opened the pull request:
###### Motivation for this change

OpenPrinting provides about 10,000 ppd files for printers, which currently need to be downloaded and installed manually for NixOS.  It would be nice to have them accessible as a package so they can be used declaratively (e.g. via `config.hardware.printers.ensurePrinters`).  This pull-request presents the necessary packages that generate and provide those ppd files.  In the course of doing so, it also adds a new setup hook that helps patching executable paths in pdd files.  The setup hook is also useful for other packages that generate ppd files -- this pull-request adds the new setup hook to those packages as well.  To keep everything clearly laid out, the changes are split into several commits.

###### Summary of the changes

Each of the following points is accomplished by one or more commits.  The commit messages contain more information on the changes and some reasoning on implementation details.

* New setup hook `patch-ppd-files`

  Many ppd files (from `foomatic-db*` as well as from other packages) contain calls to executables (e.g. `gs` or `perl`).  If these are called without absolute path, they work only if the command is in CUPS' PATH.  Other ppd files contain absolute paths (e.g. `/usr/bin/gs`) and require patching to work with NixOS at all.  Some packages already patch their ppd files to replace these commands with absolute paths into the nix store.  The setup hook `patch-ppd-files` tries to unify these efforts in a most robust way.

* Use `patch-ppd-files` on existing packages

  Some packages in nixpkgs patch executable paths in ppd files manually, e.g. with `sed`.  This pull requests replaces these efforts with the new setup hook `patch-ppd-files`.  Since I can't test those packages with real hardware (due to lack of printers), I decided to touch only those packages that already do patching so I can easily verify that results are the same and minimize the probability of breakage.

  Affected packages are

    * `cups-drv-rastertosag-gdi`
        * maintained by myself
        * the build recipe didn't patch paths before
        * also added gzip compression to ppd files
          (separate commit)
        * tested with real printer (see below)

    * `cups-kyocera`
        * maintained by @vanzef
        * just replaced patching logic with new setup hook
        * the ppd files are identical (up to package's hash)

    * `samsung-unified-linux-driver`
        * maintained by @jokogr
        * replaced patching logic with new setup hook
        * also added a missing executable's
          name to the patching logic

  The package `cups-kyodialog3` would also be a candidate, but it fails to build so I cannot verify the result.

* Introduce packages containing the Foomatic database

  The packages `foomatic-db` and `foomatic-db-nonfree` provide XML files containing the so-called Foomatic database, which contains structured information about all printers listed on https://openprinting.org/printers .  This database can be used to generate ppd files (see below).  Besides, these packages have about 6,300 ppd files on their own.  These ppd files are patched with the setup hook (see above).

* Introduce package `foomatic-db-engine`

  The package `foomatic-db-engine` contains a perl module and some command line scripts that can be used to query the Fooomatic database XML files.  The package will be used to extract ppd files.

* Introduce package `foomatic-db-ppds` containing all ppd files from the Foomatic database

  This package does not fetch a source file.  Instead, it uses the Foomatic database packages and the Foomatic database engine (see above) to generate all ppd files known to the database.



###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS (on `x86-64`)
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Tests with hardware (note that these tests were performed with stable NixOS releases as my access to printers is limited):

- [x] tested `foomatic-db-ppds/NEC/NEC-PinWriter_P6-necp6.ppd.gz` with printer "NEC PinWriter P5XL" (in NixOS 20.09, 21.05)
- [x] tested `foomatic-db-ppds/Xerox/Xerox-WorkCentre_7345-pxlcolor.ppd.gz` with printer "Xerox WorkCentre 7835 v1 Multifunction System" (in NIixOS 21.05)
- [x] tested `foomatic-db-ppds/Samsung/Samsung-ML-1440-gdi.ppd.gz` with printer "Samsung ML-1440" (in NixOS 21.05)
- [x] tested `foomatic-db-ppds/Samsung/Samsung-ML-371x-Postscript-Samsung.ppd.gz` with printer "Samsung ML-3710ND" (in NixOS 21.05)
- [x] tested `foomatic-db/Samsung/PS/Samsung_ML-371x_Series.ppd.gz` with printer "Samsung ML-3710ND" (in NixOS 21.05)
- [x] tested package `cups-drv-rastertosag-gdi` with printer `Philips 6050W 4-in-1` (in NixOS 20.09, 21.05)


###### Further comments

* Notifying `cups-kyocera` maintainer @vanzef and `samsung-unified-linux-driver` maintainer @jokogr:  This pull request replaces the ppd file patching code in your packages with the new `patch-ppd-files` hook.  As far as I can judge, the result is the same but you might want to test these packages or otherwise comment on the changes.  If you can spare the time, I would also appreciate a review of the pull request as a whole.

* Overview of similar packages in other distributions:  https://repology.org/projects/?search=foomatic-db

Yarny0 committed:
patchPpdFilesHook: new setup hook for absolute executable paths

PostScript Printer Description (ppd) files
describe printer features and capabilities.
They are usually evaluated by CUPS to convert
print jobs into a format suitable for a printer.
The conversion is often accomplished by commands
or even short shell scripts inside the ppd files.

ppd files are included in many printer driver packages.
Their scripts sometimes refer to other executables;
some of them are more common (like `perl`),
others are more exotic (like `rastertohp`).
If an executable is called with its name alone,
the effects of the ppd file depend on whether
the executable is in the PATH of CUPS,
and on the executable's version.
If an executable is called with an absolut path
(like `/usr/bin/perl`), it won't work at all in NixOS.

The commit at hand adds a setup hook that uses
the `fixupPhase` to substitute certain executable's
invocations in pdd files with absolute paths.
To use it, add the hook to `nativeBuildInputs` and
provide a list of executable names in `ppdFileCommands`.
Each executable must be available in the
package that is being built, or in `buildInputs`.
The setup hook's script then looks for ppd files in
`share/cups/model` and `share/ppds` in each output,
and replaces executable names with their absolute paths.
If ppd files need to be patched in unorthodox locations or
the setup hook needs to be invoked manually for other reasons,
one may leave the list `ppdFileCommands` empty to
avoid automatic processing of ppd files, then call
the shell function `patchPpdFileCommands` directly.
Details are described in the file `patch-ppd-hook.sh`.

Notes on the motivation for this setup hook:

Most packages in nixpkgs that provide
ppd files do not patch those ppd files at all.
This is not fatal when the executables are just called
with their names since the user can add packages
with the executables to `services.printing.drivers`.
E.g. if the user adds `pkgs.perl`, then all ppd
files that invoke `perl` will work as expected.
Nevertheless, to make these ppd files independent of
their execution environment, command invocations should
be substituted with absolut paths into the nix store.
This is similar to patching shebang lines so scripts can be
called independently of having the interpreter in the PATH.

The hook script in this commit is meant to support new packages
`foomatic-db*` which will generate several thousands of
ppd files referencing a plethora of different executables.
During development of these packages, I realized that
it's quite hard to patch ppd files in a robust way.
While binary names like `rastertokpsl` seem to be sufficiently
unique to be patched with `sed`, names like `date` or `gs`
are hard to patch without producing "false positives",
i.e., coincidental occurences of the executable's name that do
*not* refer to the executable and should not be patched at all.
As this problem also affects other packages,
it seems reasonable to put a robust implementation
in its own setup hook so that other
packages can use it without much effort.

Notes on the implementation:

The ppd file format is far from trivial.
The basic structure are key-value pairs;
keys may occur multiple times.
Only a small subset of keys may contain
executable names or shell scripts in their values.
Some values may span multiple lines;
a linebreak might even occur in the middle of a token.
Some executable names also occur in other keys by accident
where they must not be substituted (e.g. `gs` or `date`).

It is necessary to provide the list of command
names that will be patched for two reasons:
ppd files often contain "tokens" that might look
like commands (e.g. "file" or "host") but aren't;
these would erroneously get patched.
Also, looking for everything that might be a command
would slow down the patching process considerably.

The implementation uses `awk` to detect
keys that might contain executable names;
only their values are treated for substitution.
This avoids most cases of "overzealous" substitutions.
Since values may span multiple lines,
`sed` alone (while faster than `awk`) cannot focus
its substitution capabilities on relevant keys.
An elaborate set of regular expressions further helps
to minimize the probability of "false positives".
Several tricks are employed to speed up `awk`.
Notably, relevant files are identified with
`grep` before `awk` is applied to those files only.
Note that the script probably cannot handle fancy command
names (like spaces or backslashes as part of the name).
Also, there are still edge cases that the script would
mistakenly skip, e.g. if a shell script contains a
line break in the middle of an executable's name;
although ppd files permit such constellations,
I have yet to see one.

ppd files may be gzipped.
The setup hook accepts gzipped ppd files:
It decompresses them, substitutes paths, then recompresses them.
However, Nix cannot detect substituted paths as
runtime dependencies in compressed ppd files.
To ensure substituted paths are propagated as
runtime dependencies, the script adds each substituted
path to the variable `propagatedBuildInputs`.
Since this might not be enough for multi-output packages,
those paths are also written directly to
`nix-support/propagated-build-inputs`.
See the comment in `patch-ppd-hook.sh` for details.

Finally, the setup hook comes with a small test that
probes some edge cases with an artificial ppd file.

References:

* https://www.cups.org/doc/spec-ppd.html
  * general ppd file specification
  * lists some keys that may contain
    executable names or shell scripts

* https://refspecs.linuxfoundation.org/LSB_4.0.0/LSB-Printing/LSB-Printing/ppdext.html
  * lists some keys that may contain
    executable names or shell scripts

* https://en.wikipedia.org/wiki/PostScript_Printer_Description#CUPS
  * lists the usual locations of ppd files

Yarny0 committed:
cups-drv-rastertosag-gdi (cups driver): gzip ppd files

ppd files are gzipped to save some space.
The `gzip` "-n" option prevents gzip from storing
a timestamp, thus facilitating reproducibility.

Yarny0 committed:
cups-drv-rastertosag-gdi (cups driver): patch ppd bin paths

The `cups-drv-rastertosag-gdi` CUPS printer driver
package provides a filter executable `rastertosag-gdi`
and two ppd files that reference the filter file.
The commit at hand uses `patchPpdFilesHook` to
replace calls to the filter executable with
absolute paths to increase package purity.

Yarny0 committed:
foomatic-db (cups ppd files): init at unstable/2022-10-03

`foomatic-db` contains the collected knowledge about printers,
drivers, and driver options from OpenPrinting in xml files.
It also provides thousands of ppd files.
The build process patches those files to reference executables
in the nix store and gzips them to reduce storage
(from about 670 MiB to 90 MiB installed).

In contrast to the sister package `foomatic-db-nonfree`
(packaged in a follow-up commit),
this package only includes files published under a free license.
Most files are published under the GPL,
some under the MIT license.
For details see
https://github.com/OpenPrinting/foomatic-db/blob/master/COPYING .
To encompass the different licenses, the package's
license is set to "free" as the common denominator.

ppd files can be found in `$out/share/cups/model/foomatic-db`.
The subdirectory `foomatic-db` is used to avoid conflicts
with other packages that might be combined with the package
`foomatic-db` in `config.services.printing.drivers`.

There is a daily snapshot of a source archive available at
https://www.openprinting.org/download/foomatic/ .
However, these files rotate daily and
cannot be used as a stable download source.
So we rely on OpenPrinting's Github repository
instead and pinpoint a fresh commit.

Yarny0 committed:
foomatic-db-nonfree (cups ppd files): init at unstable/2015-06-05

`foomatic-db-nonfree` contains -- similar to its sister
package `foomatic-db` -- knowledge about printers,
drivers, and driver options from OpenPrinting in xml files.
It needs to be combined with the `foomatic-db`
package to yield a working database.
It also provides about 100 (gzipped) ppd files.
In contrast to `foomatic-db`, this package contains
data that is provided under non-free licenses.

Quoting
https://github.com/OpenPrinting/foomatic-db-nonfree/blob/master/README

> This is a repository of PPD and Foomatic XML files that may
> have restrictions that keep them from being used on a variety
> of machines for licensing and other non-technical reasons.

ppd files can be found in
`$out/share/cups/model/foomatic-db-nonfree`.
The subdirectory `foomatic-db-nofree` is used to avoid conflicts
with other packages that might be combined with the package
`foomatic-db-nonfree` in `config.services.printing.drivers`.

ppd files in this package are *not* patched
to call executables from the nix store:
The only executable "rastertophaser6100" that is
called from ppd files isn't available in nixpkgs.

There is a daily snapshot of a source archive available at
https://www.openprinting.org/download/foomatic/ .
However, these files rotate daily and
cannot be used as a stable download source.
So we rely on OpenPrinting's Github repository
instead and pinpoint a fresh commit.
Note that the current version is from 2015,
so updates are unlikely.

Yarny0 committed:
foomatic-db-engine: init at unstable-2022-05-03

`foomatic-db-engine` contains several perl scripts to parse
and process XML files from the Foomatic database packages.
It can be used to extract ppd files,
which will be accomplished in a follow-up commit.

The package also contains scripts
to handle print queues and jobs.
It can -- optionally -- talk to the local cups server,
to network printers and to SMB print servers.
The build recipe contains switches to enable these features;
however, they are not needed when generating ppd files.

There is a daily snapshot of a source archive available at
https://www.openprinting.org/download/foomatic/ .
However, these files rotate daily and
cannot be used as a stable download source.
So we rely on OpenPrinting's Github repository
instead and pinpoint a fresh commit.

Yarny0 committed:
cups-kyocera (cups driver): use patchPpdFilesHook

The `sed` script in the `installPhase` is removed.
Instead, the setup hook `patchPpdFilesHook`
is used to patch the path to `rastertokpsl`.

The result should essentially be the same.
Comparing the generated ppd files showed no difference
(short of the package's hash in the absolute paths).
The new package also contains a `propagated-build-inputs`
file which propagates the package itself.
This ensures the package is available whenever
a ppd file is singled out by another package.

Yarny0 committed:
samsung-unified-linux-driver (cups driver): use patchPpdFilesHook

The `sed` script in the `installPhase` is removed.
Instead, the setup hook `patchPpdFilesHook` is
used to patch the path to the filter executables.

The result should essentially be the same.
Comparing the generated ppd files showed no difference,
short of the package's hash in the absolute paths and
added newline characters at the end of some ppd files.
Missing newline characters at the end of the last line are
apparently added by `awk`; this shouldn't affect functionality.
The new package also contains a `propagated-build-inputs`
file which propagates the package itself.
This ensures the package is available whenever
a ppd file is singled out by another package.

Yarny0 committed:
foomatic-db-ppds (cups ppd files): init

`foomatic-db-ppds` uses the Foomatic database from the packages
`foomatic-db` and -- optionally -- `foomatic-db-nonfree`
and the perl modules from `foomatic-db-engine`
to generate about 8,800 ppd files.

The general structure of the build recipe is as follows:

* Merge `foomatic-db` and `foomatic-db-nonfree` into
  one package that represents the Foomatic database.
  The package `foomatic-db-nonfree` is optional
  as it taints the result license-wise;
  it will only be used if `withNonfreeDb`
  is to to `true` in the `callPackage` call.
  We create a tiny setup hook script that provides the combined
  database and sets an environment variable pointing to the
  database direcotry, which is expected by the foomatic engine.

* The final package's license and version are computed
  from the licenses and versions of the database packages.
  The license is set to `free` if each database-providing
  package has a free license, and to `unfree` otherwise.
  The version is simply the highest version
  of the database-providing packages.

* The final package uses `foomatic-compiledb`
  from the `foomatic-db-engine` package to extract
  all ppd files from the database packages.
  `patchPpdFilesHook` is used to patch most
  executable invocations in the ppd files
  so that they point to nix store paths.
  Finally, ppd files are gzipped to reduce storage
  (from about 550 MiB to 90 MiB installed).

The "nonfree" version of the package, i.e. the version that is
based on `foomatic-db-nonfree` in addition to `foomatic-db`,
contains about 120 additional ppd files
compared to the "free" version.
Since the "free" version already produces about 8,700
ppd files and hydra won't build the "nonfree" version,
the commit adds two package variables to `all-packages.nix`:

* `foomatic-db-ppds` is based on `foomatic-db` only
* `foomatic-db-ppds-withNonfreeDb`
  is also based on `foomaitc-db-nonfree`

The package introduced by this commit
is the result of combining other packages;
it is not the build product of a simple source tarball.
While it would also be possible to perform the ppd file
generation directly in the build process of the database
packages, this would yield further complexity as the
`foomatic-db-nonfree` package needs to be combined with the
`foomatic-db` package before ppd file extraction is possible.

There is no upstream product with a name that
could/should be used for the `name` attribute,
the variable name, or for the filename in nixpkgs.
Similar packages have different names across distributions:

* https://repology.org/projects/?search=openprinting
* https://repology.org/projects/?search=foomatic

The name `foomatic-db-ppds` seems to be most common
(albeit not really *that* common):

* https://repology.org/project/foomatic-db-ppds/versions

At least openSUSE splits their corresponding
package into multiple "binary" packages
(similar to our multi-output packages):

* https://build.opensuse.org/package/binaries/Printing/OpenPrintingPPDs/openSUSE_Tumbleweed

I considered something similar.
However, after doing some statistics,
I concluded that it's not worth the effort:
The biggest dependencies (`perl` and `cups-filters`) are
already present on most NixOS systems, and they cannot
be "split away" easily since it cannot be
done along a canonical line (e.g. printer driver).
Splitting directly by dependency risks that ppd files
unexpectedly "move from output to output" on package updates;
disappearing ppd files can be quite annoying for package users.

Yarny0 committed:
samsung-unified-linux-driver (cups driver): patch all filters

ppd files in the `samsung-unified-linux-driver`
package invoke these filter commands:

* pstosecps
* pstospl
* rastertospl
* pstosplc

The paths to all of those commands excluding the last one got
patched with their absolute paths during the build process.
This commit adds the last one to
the list of commands to be patched.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 133537` at 123e9175 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>cups-drv-rastertosag-gdi</li>
    <li>foomatic-db-engine</li>
    <li>samsung-unified-linux-driver (samsung-unified-linux-driver_4_01_17)</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/misc/cups/drivers/samsung/4.01.17.nix:42:3:
```
   |
42 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db/default.nix:
```diff
@@ -0,0 +1,103 @@
+
+  # compress ppd files
+  postFixup = ''
+    >&2 printf 'compressing ppd files\n'
```
```suggestion
    echo 'compressing ppd files'
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db/default.nix:
```diff
@@ -0,0 +1,103 @@
+
+  # don't let the intaller gzip ppd files as we would
+  # have to unzip them later in order to patch them
+  configureFlags = [ ''--disable-gzip-ppds'' ];
```
```suggestion
  configureFlags = [ "--disable-gzip-ppds" ];
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:
```diff
@@ -0,0 +1,136 @@
+
+  # compress ppd files
+  postFixup = ''
+    >&2 printf 'compressing ppd files\n'
```
```suggestion
    echo 'compressing ppd files'
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:
```diff
@@ -0,0 +1,136 @@
+
+in
+
+
+stdenv.mkDerivation {
+
+  pname = "foomatic-db-ppds";
```
```suggestion
stdenv.mkDerivation {
  pname = "foomatic-db-ppds";
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:
```diff
@@ -0,0 +1,136 @@
+    pathsToLink = [ "/share/foomatic" ];
+    postBuild = ''
+      mkdir -vp "${placeholder "out"}/nix-support"
+      printf 'export FOOMATICDB="${placeholder "out"}/share/foomatic"\n'  \
```
```suggestion
      echo 'export FOOMATICDB="${placeholder "out"}/share/foomatic"' \
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:
```diff
@@ -0,0 +1,136 @@
+, perl
+, psutils
+, patch-ppd-files
+}:
+
+
+let
+
+  foomatic-db-packages = [ foomatic-db foomatic-db-nonfree ];
+
+  foomatic-db-combined = buildEnv {
```
```suggestion
}:

let
  foomatic-db-packages = [ foomatic-db foomatic-db-nonfree ];

  foomatic-db-combined = buildEnv {
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db-nonfree/default.nix:
```diff
@@ -0,0 +1,87 @@
+
+  # compress ppd files
+  postFixup = ''
+    >&2 printf 'compressing ppd files\n'
```
```suggestion
    echo 'compressing ppd files'
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db-nonfree/default.nix:
```diff
@@ -0,0 +1,87 @@
+  postInstall = ''
+    test -d "${placeholder "out"}/share/foomatic/db/source/PPD"
+    mkdir -vp "${placeholder "out"}/share/cups/model"
+    ln -vs  \
+      "${placeholder "out"}/share/foomatic/db/source/PPD"  \
+      "${placeholder "out"}/share/cups/model/foomatic-db-nonfree"
```
```suggestion
    ln -vs "${placeholder "out"}/share/foomatic/db/source/PPD"  \
      "${placeholder "out"}/share/cups/model/foomatic-db-nonfree"
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/foomatic-db-engine/default.nix:
```diff
@@ -0,0 +1,75 @@
+    "PERLPREFIX=${placeholder "out"}"
+  ];
+
+  preInstall = ''mkdir -vp "${placeholder "out"}/lib/cups"'';
```
```suggestion
  preInstall = ''
    mkdir -vp "${placeholder "out"}/lib/cups"
  '';
```

SuperSandro2000 commented on pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh:
```diff
@@ -0,0 +1,202 @@
+        # create list of affected files
+        xargs -0r  grep -lZF "$binnew"  < ppds  > patched-ppds  ||  true
+
+        >&2 printf 'patchPpdFileCommands: %s ppd file(s) patched with '"'%s'"'\n'  \
```
```suggestion
        echo 'patchPpdFileCommands: %s ppd file(s) patched with '"'%s'"''  \
```

SuperSandro2000 commented on pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh:
```diff
@@ -0,0 +1,202 @@
+# shell or regex escaping (e.g. backslash).
+
+function patchPpdFileCommands () {
+
+    local bin
+    local binnew
+    local binold
+    local binoldgrep
+    local cupspath
+    local path
+    local ppdroot
+    local ppdrootprefix
```
```suggestion
    local bin binnew binold binoldgrep cupspath path ppdroot ppdrootprefix
```

SuperSandro2000 reviewed (commented):
You get the gist. I did not comment everything.
Also we should probably remove the -v from most commands to make the build less noisy.

Yarny0 committed:
cups-kyocera (cups driver): call {pre,post}Install

Add `preInstall` and `postInstall` to the `installPhase` to
permit package overrides to add commands to the install phase.

Suggested by r-rmcgibbo bot in
https://github.com/NixOS/nixpkgs/pull/133537#issuecomment-897437621

Yarny0 committed:
samsung-unified-linux-driver (cups driver): call {pre,post}Install

Add `preInstall` and `postInstall` to the `installPhase` to
permit package overrides to add commands to the install phase.

Suggested by r-rmcgibbo bot in
https://github.com/NixOS/nixpkgs/pull/133537#issuecomment-897437621

SuperSandro2000 commented:
> I'm not sure if all echo implementations that might come into play here can handle situations where parts of the string look like a command line option.

Luckily that does not matter because always the bash from the stdenv is used.

Yarny0 commented:
Thanks @SuperSandro2000.
The branch is updated, all shell script messages are now performed with `echo`.  It is also rebased on top of current `nixos-unstable` commit.
I don't expect any functional changes or breakage due to the update, as it only affects diagnostic messages.  However, to rule out typos I rebuilt all affected packages (on NixOS unstable) and quickly tested the package again, with `foomatic-db-ppds/NEC/NEC-PinWriter_P6-necp6.ppd.gz` with printer "NEC PinWriter P5XL" (in NixOS 21.05), and the printer still prints as expected.

Yarny0 commented:
Another update of the pull request's branch as I just realized https://github.com/OpenPrinting/foomatic-db received some commits in August and September.
They added xml/ppd files for five Ricoh printers:  https://github.com/OpenPrinting/foomatic-db/commit/13e1f23b4dae7987921b89a07fd5ddb0ae5c7a51#diff-91c5b46dc84a94604a4e4d0caed9bf85590a2eddbb12d2e8dc80badf324a9dfb
All packages in the pull request at hand still build fine with the update.  The new ppd files do not contain any new commands that would require an update of the list of commands to be patched.

SuperSandro2000 commented on pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh:
```diff
@@ -0,0 +1,192 @@
+    local bin binnew binold binoldgrep cupspath path ppdroot ppdrootprefix
+
+    # we will store some temporary data here
+    pushd "$(mktemp -d --tmpdir patch-ppd-file-commands.XXXX)"
```
Right, if this only exists in the sandbox then this is fine.

SuperSandro2000 commented:
@ofborg eval
@ofborg build

Yarny0 committed:
cups-drv-rastertosag-gdi (cups driver): fix simple comment typo

Yarny0 commented:
Branch updated:

* updated `foomatic-db` to version `2022-01-27`
* rebased onto fresh `nixos-unstable`

Yarny0 commented:
Pushed another update:

* updated `foomatic-db` to version `2022-03-27`
* rebased onto current `nixos-unstable` (e10da1c7f542515b609f8dfbcf788f3d85b14936)

All affected packages still build fine.

Tested `foomatic-db-ppds/Samsung/Samsung-ML-371x-Postscript-Samsung.ppd.gz` with printer "Samsung ML-3710ND" (with current NixOS unstable and with current NixOS 21.11).

Is there anything that can be done to help this pull request getting merged into `master`?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/911

Yarny0 commented:
Another update of the branch.

* Rebase onto current `nixos-unstable`, i.e., 93c57a988470c1948976b1bb70abbd5855c5b810 .
* Moved the string "unstable" from the `pname` to the `version` attributes, following 01ec4349f20157dd3bbcc1f3a540eb051c47be50 .
* All new packages still refer to the current version's sources.
* All new/affected packages still build fine, up to one package: The unfree package `cups-kyocera` fails to download it's source archive; the server apparently no longer provides the file.  Notifying maintainer: @vanzef .  I'm still retaining the `cups-kyocera` related commits as they worked fine before and will likely work fine as soon as the package's source archive is available again.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1070

Yarny0 commented:
Updated and rebased the branch.  The branch is now using the newest version of `foomatic-db`, released yesterday.  The release added a couple of RICOH printers and an epson printer.
                                                                                                                                    
All packages still build fine, and the automated test for the new `patch-ppd-files` setup hook still passes.                        
                                                                                                                                    
It would really be nice to have this pull request finally be merged into master.

Yarny0 commented:
Another rebase, including the newest version of `foomatic-db`.  Everything still builds fine.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/52

doronbehar commented on pkgs/misc/cups/drivers/foomatic-db-ppds/foomatic-ppdfiles.pl:
```diff
@@ -0,0 +1,82 @@
+# be used to extract ppd files from the database.
+# However, that script only produces one ppd file per call,
+# but it needs to process the complete database to do so.
+# To accelerate the process (about two orders of magnitude),
```
Have you considered contributing this script upstream? With or without this script part of `foomatic-db-engine`, I envision that `foomatic-db-ppds` could be removed, and you'll run that script in `foomatic-db{,-nonfree}` (perhaps conditionally) - making these packages include the ppd files from the databases they came with.

Yarny0 replied:
Thanks for bringing up this idea.  The `foomatic-db-engine` maintainer [answered promptly](https://github.com/OpenPrinting/foomatic-db-engine/pull/8#issuecomment-1296183602) and brought another command to my attention:  I didn't spot `foomatic-compiledb` earlier, but it is doing exactly what I tried to accomplish with my tiny script.  Up to the subdirectory structure, the result is the same, so we don't have to patch `foomatic-db-engine` at all.

Note: I also upgraded the `foomatic-db-engine` from 4.0 to 4.1 while preparing the patch.  With this update, the final `foomatic-db-ppds` package "lost" about 1.000 PPD files, but apparently these files were duplicates of files in `foomatic-db`, so we don't really loose anything.

Regarding the ppd export directly in the database packages: Inspired by your idea, I restructured everything to make this work, but alas, `foomatic-db-nonfree` isn't a "database" by itself, but just a kind of extension to the free `foomatic-db`.  The free database would install and export everything alright, but the nonfree version would need to be combined with the free one after installation, only to generate the free ppd files again.  The result of this restructuring would produce working packages, but the build script of the nonfree version would work very differently (and more complicated) compared to the free version, leading to a quite awkward structure overall.  Finally, I came to the conclusion that the current design is more understandable and maintainable, so I suggest to keep it.

doronbehar replied:
The current packages' structure is great, I'm glad you switched to `foomatic-compiledb`. Thanks for the effort!

doronbehar commented on pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:
```diff
@@ -0,0 +1,128 @@
+    name = "foomatic-db-combined";
+    paths = foomatic-db-packages;
+    pathsToLink = [ "/share/foomatic" ];
+    postBuild = ''
```
If we'll leave things as they are regarding this new package, I think it'd be nice to add a comment like this (assuming I understood what's going on correctly):

```suggestion
    # `foomatic-db-combined` is a nativeBuildInput of `foomatic-db-ppds`. Thanks to this
    # setup hook, our `foomatic-ppdfiles.pl` script knows where to find the database to
    # generate ppd files from.
```

Yarny0 replied:
Yes, I appended a corresponding comment.

doronbehar reviewed (commented):
Everything is very well written, and commit messages are very detailed! Thanks for the effort to contribute.

Yarny0 commented on pkgs/misc/cups/drivers/cups-drv-rastertosag-gdi/default.nix:
```diff
@@ -35,6 +35,9 @@ python3Packages.buildPythonApplication rec {
     runHook postInstall
   '';
+  postFixup = ''
+    gzip -9nv "${placeholder "out"}/share/cups/model/rastertosag-gdi"/*.ppd
```
Yeah, that sounds reasonable.  I added a hint in the release notes and will post a warning in discourse after it got merged.

doronbehar replied:
> Yeah, that sounds reasonable. I added a hint in the release notes and will post a warning in discourse after it got merged.

Just a friendly reminder :)

Yarny0 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1064,6 +1064,8 @@ with pkgs;
     { name = "validate-pkg-config"; deps = [ findutils pkg-config ]; }
     ../build-support/setup-hooks/validate-pkg-config.sh;
 
+  patch-ppd-files = callPackage ../build-support/setup-hooks/patch-ppd-files {};
```
I agree.  Adding `Hook` in hooks' names seems helpful in general.  Done.

Yarny0 commented on pkgs/misc/cups/drivers/samsung/4.01.17.nix:
```diff
@@ -74,7 +74,12 @@ in stdenv.mkDerivation rec {
+    "pstosecps"
+    "pstospl"
+    "pstosplc"
+    "rastertospl"
```
I usually try to separate commits if their changes are conceptually different, to facilitate bisection if a bug is introduced.  Here, the first commit should not affect the package's output at all (besides its hash), while the second commit intentionally alters the output, as it fixes an oversight that was already present for a long time.  I think the separation into two commits is justified here, what do you think?

doronbehar replied:
> I usually try to separate commits if their changes are conceptually different, to facilitate bisection if a bug is introduced. [...] I think the separation into two commits is justified here, what do you think?

I agree with this justification for separating / merging commits. Still I would have merged these two commits because I wouldn't have had the patience to vouch for the previous change to not change the resulting ppd files, and only change the inputs. But that's me, you are much more thorough then me! :)

Yarny0 commented on pkgs/misc/cups/drivers/foomatic-db/default.nix:
```diff
@@ -0,0 +1,96 @@
+  # use a package-specific subdirectory to avoid
+  # conflicts with other ppd-containing packages
+  postInstall = ''
+    [[ -d "${placeholder "out"}/share/foomatic/db/source/PPD" ]]
```
Yes, that's the idea of this line.  I expanded the line to print an error message.

doronbehar commented on pkgs/misc/cups/drivers/foomatic-db-engine/default.nix:
```diff
@@ -0,0 +1,81 @@
+  # The following package would be "found"
+  # by the configure script if available,
+  # it might improve something in the final package:
+  # a2ps ghostscript
```
What do you mean by "improve something"? You noticed the configure script looks for these commands, but you didn't understand what's the usage of the configure script / make files with these commands?

Yarny0 replied:
Somehow yes.  If you put `a2ps` into `nativeBuildInputs`, the configure script will say something like

```
checking for a2ps... /nix/store/fhp0m7anlcx4zpzz7qy0kkyilpag5j5n-a2ps-4.14/bin/a2ps
```

For `a2ps` specifically, I couln't find any usage inside the `foomatic-db-engine` sources, so it *might* be some kind of leftover, but I can't really be sure.

This is similar to the list I put in another comment a few lines above: There are also packages that are referenced by the configure script if found, but they will also become a runtime dependency of the resulting `foomatic-db-engine` package in the end (in contrast to `a2ps` and `ghostscript` which will be "used" during the build process, but not be referenced by the result).

The result works fine without any of those packages as build inputs.  On the other hand, the configure script demands certain build inputs (like the `file` utility), so I came to the conclusion that these packages are deliberately treated as optional inputs and can be omitted to avoid bloating the build and the closure.

However, someday someone might be missing a feature that could be unlocked with one of those optional input packages and I hope these comments might point that someone into the right direction and save some debugging time.

doronbehar replied:
`./configure` scripts looking up for `file` is a known harmless issue happening with all autotools based software, nothing I'd worry about.

OTH, `foomatic-db-engine`'s README explains what is the usage of the `a2ps` and ghostscript and it seems to me, that they can be used during runtime as filters or something like that. It's not a reason to add them to `nativeBuildInputs`. As for the `buildInputs`, it would have been nice to track down what are the new features gained by adding `cups`, `cups-filter` etc.

If I understand correctly, that most of `foomatic-db-engine`'s purpose is to help build `foomatic-db-ppds`, I trust you that most users will likely use the ppds and that the generated ppds are good. Still, perhaps it'd be best to add these inputs conditionally, defaulting to not adding them; and ideally, explain better in a comment above the `lib.optional` lines what is the feature gained by adding these dependencies. Perhaps permalinks to upstream's README is best.

Yarny0 replied:
Sadly, the README file didn't really help here.  It recommends to have a couple of additional programs installed, but some of those are not really used by the engine package (probably just recommended to make certain PPD files work).  Other dependencies are not mentioned in the README.

I tracked down the references in the perl code: `ghostscript` and `a2ps` are not used at all; they do not alter the output of the package in any way.  The other dependencies expand the functionalities of the scripts `foomatic-{configure,getpjloptions,printjob}`.  From what I understand, these script can be used to modify printer configuration (e.g. add CUPS queues) or submit print jobs, so they are indeed not used in the process of PPD file generation.

I added three switches `withCupsAccess`, `withSocketAccess` and `withSMBAccess` that control the addition of optional dependencies, as you suggested.  I was thinking about the common name `with...Support` that is used elsewhere in nixpkgs, but concluded it might confuse other contributors into thinking that `foomatic-db-engine` or the PPD files it generates cannot be used with cups unless `withCupsSupport` is enabled.  The name `with...Access` makes it more clear that this is only needed if the resulting package is going to be used to *access* one of those entites directly.

Those new switches are the result of my interpretation of the perl code (in particular, the `sysdeps` structure created by `makeDefaults.pm`).  I'm quite sure that access to cups, to network socket and to SMB can be separated the way presented here.  It might even be possible to divide the cups access feature up further in a meaningful way, but from looking at the code, I can't spot an obvious way to do this.

doronbehar replied:
Very good job on `foomatic-db-engine`! Thanks again!

doronbehar commented:
Result of `nixpkgs-review pr 133537` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>cups-kyocera</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>cups-drv-rastertosag-gdi</li>
    <li>foomatic-db</li>
    <li>foomatic-db-engine</li>
    <li>foomatic-db-nonfree</li>
    <li>foomatic-db-ppds</li>
    <li>foomatic-db-ppds-withNonfreeDb</li>
    <li>patchPpdFilesHook</li>
    <li>samsung-unified-linux-driver (samsung-unified-linux-driver_4_01_17)</li>
    <li>samsung-unified-linux-driver_1_00_36</li>
    <li>samsung-unified-linux-driver_1_00_37</li>
  </ul>
</details>

doronbehar commented:
`cups-kyocera` failing to build is not an issue of this PR - the URL of the source seems a bit flaky. Not should be a blocker for this well written PR.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/15
